lib/rubocop/cop/lint/next_without_accumulator.rb

```diff
@@ -0,0 +1,50 @@
+# encoding: utf-8
+
+module RuboCop
+  module Cop
+    module Lint
+      # Don't omit the accumulator when calling `next` in a `reduce` block.
+      #
+      # @example
+      #   # bad
+      #   result = (1..4).reduce(0) do |acc, i|
+      #     next if i.odd?
+      #     acc + i
+      #   end
+      #
+      #   # good
+      #   result = (1..4).reduce(0) do |acc, i|
+      #     next acc if i.odd?
+      #     acc + i
+      #   end
+      class NextWithoutAccumulator < Cop
+        MSG = 'Use `next` with an accumulator argument in a `reduce`.'.freeze
+        METHODS = [:inject, :reduce]
+
+        def on_block(node)
+          method, _args, body = *node
+
+          # filter out super and zsuper nodes
+          return unless method.type == :send
+
+          _, method_name, method_args = *method
+          return unless METHODS.include? method_name
+          return if method_args && method_args.type == :sym
+
+          return false unless body.type == :begin
+
+          # Astrolabe::Node#each_node FTW!
+          # (actually we want a recursive detect
+          # but astrolabe cannot do a short circuit on a match)
+          found_void_next = nil
```

Comment:
This is not necessary. Once you find a "void next" node, you can call `add_offense` and then just `return`.